ml-prims/src/hmm/magma/b_bilinear.h

```diff
@@ -0,0 +1,109 @@
+#pragma once
+
+#include "cublas_v2.h"     // if you need CUBLAS v2, include before magma.h
+// #include "magma.h"
+// #include "magma_lapack.h"  // if you need BLAS & LAPACK
+
+#include "hmm/magma/magma_test_utils.h"
+#include "hmm/magma/magma_batched_wrappers.h"
+// #include "cuda_utils.h"
+
+using namespace MLCommon::LinAlg;
+
+namespace MLCommon {
+
+template <typename T>
+__host__ __device__
+T dot(int n, T* x, T* y){
+        T res = 0;
+        for (size_t i = 0; i < n; i++) {
+                res += x[i] * y[i];
+        }
+        return res;
+}
+
+template <typename T>
+__global__
+void dot_batched_kernel(int n, T **dX_array, T **dY_array, T *dO,
+                        magma_int_t batchCount, int numThreads){
+        int idxThread = threadIdx.x + blockDim.x * blockIdx.x;
+        for (size_t i = idxThread; i < batchCount; i+=numThreads) {
+                dO[i] = dot(n, dX_array[i], dY_array[i]);
+        }
+}
+
+template <typename T>
+void dot_batched(int n, T **dX_array, T **dY_array, T *dO,
+                 magma_int_t batchCount){
+        dim3 block(32, 1, 1);
+        dim3 grid(ceildiv(batchCount, (int)block.x), 1, 1);
+        int numThreads = grid.x * block.x;
+        dot_batched_kernel<T> <<< grid, block >>>(n, dX_array, dY_array, dO,
+                                                  batchCount, numThreads);
+        cudaDeviceSynchronize();
```

Comment:
Please remove this. We'd want the kernel calls to remain asynchronous for perf reasons.